### PR TITLE
GH-40113: [Go][Parquet] New RegisterCodec function

### LINKED ISSUE #40113: [Go][Parquet] Allow registering new codecs for parquet
### Describe the enhancement requested

Current Parquet implementation in Go does not support all codec defined in the standard, some due to licensing issues, which is fine. And the choice of implementation is also not necessarily the best for every use cases. For example, I would like to be able to use libdeflate for GZIP codec or CGO binding of ZSTD codec. Instead of updating the implementation in the main repo, I think it is better to allow user to supply their own. For example, CGO may not be suitable for all use cases and hence may not be suitable as the default implementation.

### Component(s)

Go, Parquet

### PR DESCRIPTION
**Pull request by zhouyan**
This is to allow addition/overwrite of custom codec implementation

This allows other modules to provide alternative implementations for the compression algorithms, such as using libdeflate for Gzip, or CGO version of ZSTD.

In addition, it allows others to supply codecs that cannot be easily supported by this library such as LZO due to license reasons or LZ4.

### Rationale for this change

See #40113

### What changes are included in this PR?

A new RegisterCodec function added 

### Are these changes tested?

yes

### Are there any user-facing changes?

It's an addition more targeted towards library writers. 

* Closes: #40113

### CODE REVIEW

File: go/parquet/compress/compress.go [hunk 1]
```diff
@@ -92,6 +92,10 @@ type Codec interface {
 
 var codecs = map[Compression]Codec{}
 
+func RegisterCodec(compression Compression, codec Codec) {
+	codecs[compression] = codec
+}
```
Comment (zeroshade) [inline go/parquet/compress/compress.go:97]: Can you add godoc comments to this?

**Review by zeroshade — CHANGES_REQUESTED** [review]
This looks good to me in general, just needs some comments for the documentation

File: go/parquet/compress/compress.go (hunk 1 above)
Comment (zhouyan) [reply go/parquet/compress/compress.go:113]: Thanks! Added some basic comment and simple usage example

File: go/parquet/compress/compress.go [hunk 2]
```diff
@@ -92,6 +92,22 @@ type Codec interface {
 
 var codecs = map[Compression]Codec{}
 
+// RegisterCodec add new or override existing codec implementation for a given compression algorithm.
```
Comment (zeroshade) [inline go/parquet/compress/compress.go:95]:
```suggestion
// RegisterCodec adds or overrides a codec implementation for a given compression algorithm.
```

File: go/parquet/compress/compress.go [hunk 3]
```diff
@@ -92,6 +92,22 @@ type Codec interface {
...
+//		    RegisterCodec(compress.Codecs.Zstd, czstdCodec{})
+//		}
+//
+//		type czstdCodec struct{} // implementing Codec interface using CGO based ZSTD wrapper
+//
+// And user of the custom codec can import the above package like below,
+//
+//	    package main
+//
+//		import _ "package/path/to/czstd"
```
Comment (zeroshade) [inline go/parquet/compress/compress.go:110]: Can we fix the spacing?

**Comment by zeroshade** [comment]
@zhouyan We squash when we merge anyways, so there's no need to manually do so yourself.

**Review by zeroshade — APPROVED** [review]

**Pull request merged**